### PR TITLE
feat(#70): adds Context Request Protocol Implementation / Dependency Injection

### PR DESCRIPTION
**Pull request by quarkus**

### CODE REVIEW

**Comment by quarkus** [comment]
@lukas-schardt do you think we can make use of these maps in luna-js ?
Maybe shim them via lunas dependency injection ?

File: src/BaseElement.js [hunk 1]
```diff
@@ -189,7 +196,7 @@ class BaseElement extends HTMLElement {
 	 * Will trigger update() when a property was changed
 	 */
 	defineProperties() {
-		Object.keys(this.properties()).forEach((prop) => {
+		Object.keys({ ...this.injectProperties(), ...this.properties() }).forEach((prop) => {
```
Comment (eddyloewen) [inline src/BaseElement.js:199]: @quarkus does the order matter here? I mean you said it was not allowed to have the same property in both maps - what what if someone did it anyways (because yolo). Would it break?

File: src/BaseElement.js [hunk 2]
```diff
@@ -294,6 +305,58 @@ class BaseElement extends HTMLElement {
...
+		if (Object.keys(this.provideProperties()).length > 0) {
+			this.addEventListener('request-context', this.onRequestContext);
+		}
+	}
+
+	requestContext(propertyName, valueOrCallback) {
+		this.dispatch('request-context', { [propertyName]: valueOrCallback }, true);
+	}
+
+	onRequestContext(event) {
```
Comment (eddyloewen) [inline src/BaseElement.js:341]: @quarkus This is still using and mixing provide/inject with context wording...

File: src/BaseElement.js (hunk 1 above)
Comment (quarkus) [reply src/BaseElement.js:199]: nope .. thats why the order matters ;) .
In this implemetation its actually ok to have it in both maps but due to the order proerty defaults from the property map.
If its missing in property map this will autodefine a prop for the current key.

File: src/BaseElement.js (hunk 2 above)
Comment (quarkus) [reply src/BaseElement.js:341]: Yes .. but thats actually intentional to use the requestContext Protocol "under the hood" and to interface it through the inject/provideProperties to the developer.

Comment (eddyloewen) [reply src/BaseElement.js:341]: Alright

File: src/BaseElement.js (hunk 1 above)
Comment (eddyloewen) [inline src/BaseElement.js:199]: @quarkus is this the correct order? Or would we expect the injected properties to override the default properties? I'm not sure what is best.

File: src/BaseElement.js [hunk 3]
```diff
@@ -238,6 +245,10 @@ class BaseElement extends HTMLElement {
 				if (JSON.stringify(oldValue) !== newValueString) {
 					this._state[property] = newValue;
 
+					if (newValue instanceof Store) {
+						newValue.subscribe(this);
+					}
+
```
Comment (eddyloewen) [inline src/BaseElement.js:251]: @quarkus this should be already merged from another branch ?!

File: src/BaseElement.js [hunk 4]
```diff
@@ -294,6 +305,59 @@ class BaseElement extends HTMLElement {
...
+	/**
+	 * Define which properties that can be provided to children
+	 * @return {{}}
+	 */
+	provideProperties() {
+		return {};
+	}
+
+	/**
+	 * Defines context on the element based on keys from this.context()
```
Comment (eddyloewen) [inline src/BaseElement.js:325]:
```suggestion
	 * Defines context on the element based on keys from this.injectProperties()
```

File: src/BaseElement.js [hunk 5]
```diff
@@ -294,6 +305,59 @@ class BaseElement extends HTMLElement {
...
+		Object.entries(this.injectProperties()).forEach(([key, value]) => {
+			this.requestContext(key, value);
+		});
+		// define context provider
+		if (Object.keys(this.provideProperties()).length > 0) {
+			this.addEventListener('request-context', this.onRequestContext);
+		}
+	}
+
+	requestContext(propertyName, valueOrCallback) {
```
Comment (eddyloewen) [inline src/BaseElement.js:337]: @quarkus can we add some JSDocs?

File: src/BaseElement.js [hunk 6]
```diff
@@ -294,6 +305,59 @@ class BaseElement extends HTMLElement {
...
+		this.dispatch('request-context', { [propertyName]: valueOrCallback }, true);
+	}
+
+	onRequestContext(event) {
```
Comment (eddyloewen) [inline src/BaseElement.js:341]: @quarkus can we add some JSDocs?

File: src/BaseElement.js (hunk 3 above)
Comment (quarkus) [reply src/BaseElement.js:251]: hmm . but its not !?

File: src/BaseElement.js (hunk 1 above)
Comment (quarkus) [reply src/BaseElement.js:199]: I think we were going back an forth with this .. 

I think the order is correct as it prioritizes the (dafault) values of the properties over the default values from the injection map. 

As Properties are more crucial to element-js we decided to not overwrite them ..

File: src/BaseElement.js (hunk 3 above)
Comment (quarkus) [reply src/BaseElement.js:251]: Ok .. i looked into it again.

Usually an element is registered as a Store observer once its properties are defined.
The  above covers the case when a nested element requests a value which is an instance of Store.

To maintain reactivity we need to register that the store is injected to as an observer on some point.

I think we were discussing where to so it in the other Branches and ended up doing it here. 
The alternative was to subscribe the element  to the store within the "onRequestContext" method.

Comment (quarkus) [reply src/BaseElement.js:251]: ... just lost another long comment. thanks github.

tldr;
Doing this in the setter has the advantage that it works without the DI Feature and would also work for runtime assignments of store instances to other components.

That is a good thing

File: src/BaseElement.js (hunk 5 above)
Comment (quarkus) [reply src/BaseElement.js:343]: yes!

File: src/BaseElement.js (hunk 6 above)
Comment (quarkus) [reply src/BaseElement.js:354]: sure.

**Review by eddyloewen — APPROVED** [review]
I like it and I might need it next week, so go 😅
